### DISCUSSION TITLE
How to make a single window not take up the whole screen. (For ultrawide monitors)

### DISCUSSION THREAD

**Discussion by trentshaines** [nikitabobko/AeroSpace#698] [questions-and-answers] [answered]
Firstly, I love AeroSpace!

However, I want dynamic gap sizing - I don't want big gaps when I have more than one window open, I only want large gaps on this monitor when only a single window is open. Specifically, on my ultra wide, I don't want a single app to cover the whole monitor.

Is there a way to do this? Even if it isn't automatic, that would still be amazing. 

I am aware its possible to assign a monitor have different gaps for left and right. 
Ex (in aerospace.toml)
```
[gaps]
outer.left = [{ monitor."Odyssey G93SC" = 500 }, 5]
outer.right = [{ monitor."Odyssey G93SC" = 500 }, 5]
```
But, this is constant, even when I have multiple windows.

I thought maybe this would be possible with binding modes, but it looks like thats to switch between keybindings. Even with keybindings and callbacks, I'm not too sure how to make this gaps changed based on # of windows. Maybe there is a way to switch to floating and then resize, but I'm not sure exactly how, and I don't know how to detect window count either.

The only way I know for sure to do this right now is the following (which seems not ideal) : 
Toggle floating window via service mode, use rectangle software to tile in the middle, and toggle back to tiling via service mode before opening another window

Thanks!

**Comment by wesbos** [comment]
I also have this question - I have a portrait monitor to my left that I use to keep email / chat / music open on. 

Ill often close chat + email when I'm focusing and just have music open - but then it snaps to the entire monitor. 

The fix right now is to make it a floating window, but it's not ideal because you give up a lot. 

I'm not sure I know the fix for this  - we need "ghost windows", that take up space. Or being able to resize a window so it doesn't take 100% width/height of it's tile.

**Reply by trentshaines** [reply]
Yep, I sometimes opened up terminals on the left and right as fake ghost windows, but its not a great solution haha. Even if we had proper 'ghost windows', It wouldn't be ideal to manage them when I open more than one window. I think if we can change the gaps setting with a hotkey that would be great!

**Comment by fcpauldiaz** [comment]
i have the same issue when using a big monitor

**Comment by pauldub** [comment]
I've been developing two features on a personal Aerospace fork to address this issue:

- **Single-window gap configuration**: allows you to define custom gaps when there's only one window in a workspace, preventing it from taking up the entire screen.

- **Configurable focus mode**: similar to the expand command, but with adjustable gaps so you can focus on a window while maintaining a comfortable, readable size rather than full-screen expansion.

Both features aim to provide better window sizing control and improve the experience on ultra wide monitors. I'd be happy to submit a patch if these features are deemed useful.

**Comment by nikitabobko** [comment, marked as answer]
Tracking issue for ultrawide monitors support: https://github.com/nikitabobko/AeroSpace/issues/60

**Comment by scorchio** [comment]
There's something I would like to point out that I haven't seen so far, I hope it will help Nikita to come up with a solution that fits everyone's needs.

I think whether an app works well when taking the full screen on ultrawide or not is largely dependent on the application itself. 

Applications that are centered around written content (e.g. browsers, chat applications), or lists (e.g. a Finder window) are usually meant to scroll vertically, and they tend to fair poorly when given too much horizontal space. Consider that according to [this article](https://www.uxpin.com/studio/blog/optimal-line-length-for-readability/), _"The ideal line length for readable text is 50–75 characters per line (CPL), with 66 CPL being the sweet spot"_, so web designers following similar guidelines will create sites that only uses less than half of the available horizontal space for the actual content. (I also need to point out that there are widescreen monitors that are not curved, and I'm using one. Poorer quality models might have viewing angles where the legibility of the content on the side edges can be still somewhat compromised, as you're looking at it with a slight angle, not straight. Because of this, I tend to favor reading on the center of the screen.)

However, there are some software that are built around horizontal scrolling in mind. Creation software focused around timelines (audio / video production software) fall into this category. For such applications, IMO it is actually preferable to let them use the full width of an ultrawide. This is definitely the case with the music production software I use, Ableton Live, where being able to use more horizontal space makes both managing the structure of songs and operating the mixer console easier.

As far as I can tell, it seems that most of the applications fall into the first category, and the latter is more of an exception. There is also a sort of a middle ground when the application has a built-in tiling solution, like IDEs. For example, if I have a workspace with a single VS Code with a single file being open, it looks kind of awkward as long as I don't open another file to the side, or until I switch to the ["Centered editor layout" mode of VS Code](https://code.visualstudio.com/docs/getstarted/userinterface#_centered-editor-layout) that was designed to mitigate this. For such applications, I would not limit their maximum width either, because in the majority of the time I use them in situations with multiple files when they need the horizontal space, and it's better to have a temporarily awkward look than to have to disable the limiting of the width once an extra file is opened.

Based on this, my ideal solution...

- would limit the maximum width of the window to some sensible default when there's only a single window on the workspace, and would apply this to the majority of the applications;
- would enable exceptions to this, so that applications that can actually utilize the extra horizontal space would be able to do so.

For the exceptions, conditions similar to the ones in `on-window-detected` callback would work very well; my needs would be covered with app id, and I can imagine some cases where window title regex could make sense.
